std/arith.asm

```diff
@@ -0,0 +1,174 @@
+    let CLK32_22 = |row| clock(22, row);
+    let CLK32_23 = |row| clock(23, row);
+    let CLK32_24 = |row| clock(24, row);
+    let CLK32_25 = |row| clock(25, row);
+    let CLK32_26 = |row| clock(26, row);
+    let CLK32_27 = |row| clock(27, row);
+    let CLK32_28 = |row| clock(28, row);
+    let CLK32_29 = |row| clock(29, row);
+    let CLK32_30 = |row| clock(30, row);
+    let CLK32_31 = |row| clock(31, row);
+    let CLK32 = [
+        CLK32_0, CLK32_1, CLK32_2, CLK32_3, CLK32_4, CLK32_5, CLK32_6, CLK32_7,
+        CLK32_8, CLK32_9, CLK32_10, CLK32_11, CLK32_12, CLK32_13, CLK32_14, CLK32_15,
+        CLK32_16, CLK32_17, CLK32_18, CLK32_19, CLK32_20, CLK32_21, CLK32_22, CLK32_23,
+        CLK32_24, CLK32_25, CLK32_26, CLK32_27, CLK32_28, CLK32_29, CLK32_30, CLK32_31
+    ];
+
+    /****
+    *
+    * LATCH POLS: x1,y1,x2,y2,y3
+    *
+    *****/
+
+    array::map(x1, |e| unchanged_until(e, CLK32[31]));
+    array::map(y1, |e| unchanged_until(e, CLK32[31]));
+    array::map(x2, |e| unchanged_until(e, CLK32[31]));
+    array::map(y2, |e| unchanged_until(e, CLK32[31]));
+    array::map(y3, |e| unchanged_until(e, CLK32[31]));
+
+    /****
+    *
+    * RANGE CHECK x1,y1,x2,y2,y3
+    *
+    *****/
+
+    sum(16, |i| x1[i] * CLK32[i]) + sum(16, |i| y1[i] * CLK32[16 + i]) in BYTE2;
+    sum(16, |i| x2[i] * CLK32[i]) in BYTE2;
+    sum(16, |i| y3[i] * CLK32[i]) + sum(16, |i| y2[i] * CLK32[16 + i]) in BYTE2;
+
+
+    /*******
+    *
+    * EQ0: A(x1) * B(y1) + C(x2) = D (y2) * 2 ** 256 + op (y3)
+    *        x1 * y1 + x2 - y2 * 2**256 - y3 = 0
+    *
+    *******/
+
+    /// returns a(0) * b(0) + ... + a(n - 1) * b(n - 1)
+    let dot_prod = |n, a, b| sum(n, |i| a(i) * b(i));
+    /// returns |n| a(0) * b(n) + ... + a(n) * b(0)
+    let product = |a, b| |n| dot_prod(n + 1, a, |i| b(n - i));
+    /// Converts array to function, extended by zeros.
+    let array_as_fun = [|arr| |i| if 0 <= i && i < array::len(arr) {
+        arr[i]
+    } else {
+        0
+    }][0];
+    let shift_right = |fn, amount| |i| fn(i - amount);
+
+    let x1f = array_as_fun(x1);
+    let y1f = array_as_fun(y1);
+    let x2f = array_as_fun(x2);
+    let y2f = array_as_fun(y2);
+    let y3f = array_as_fun(y3);
+
+    // Defined for arguments from 0 to 31 (inclusive)
+    let eq0 = (|| |nr|
+        product(x1f, y1f)(nr)
+        + x2f(nr)
+        - shift_right(y2f, 16)(nr)
+        - y3f(nr)
+    )();
+    
+    // Note that Polygon uses a single 22-Bit column. However, this approach allows for a lower degree (2**16)
+    // while still preventing overflows: The 32-bit carry gets added to 32 16-Bit values, which can't overflow
+    // the Goldilocks field.
+    pol witness carry_low, carry_high;
+    { carry_high } in { BYTE2 };
+    { carry_low } in { BYTE2 };
+
+    pol carry = carry_high * 2**16 + carry_low;
+    
+    carry * CLK32[0] = 0;
+
+    let eq0_sum = sum(32, |i| eq0(i) * CLK32[i]);
+    carry + eq0_sum = carry' * 2**16;
```

Comment:
This means that the carry needs to be zero at the end, and this is why we have D / y2 to take the actual carry, right?

Reply:
Yes! And the other equations also should not have a carry in the end, because this should have been absorbed in the quotient.